src/screen-cast.c

```diff
@@ -0,0 +1,1050 @@
+                                     G_DBUS_PROXY_FLAGS_NONE,
+                                     g_dbus_proxy_get_name (G_DBUS_PROXY (impl)),
+                                     request->id,
+                                     NULL, &error);
+  if (!impl_request)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  request_set_impl_request (request, impl_request);
+  request_export (request, g_dbus_method_invocation_get_connection (invocation));
+
+  g_variant_builder_init (&options_builder, G_VARIANT_TYPE_VARDICT);
+  options = g_variant_builder_end (&options_builder);
+
+  g_object_set_qdata_full (G_OBJECT (request),
+                           quark_request_session,
+                           g_object_ref (session),
+                           g_object_unref);
+  screen_cast_session->state = SCREEN_CAST_SESSION_STATE_STARTING;
+
+  xdp_impl_screen_cast_call_start (impl,
+                                   request->id,
+                                   arg_session_handle,
+                                   request->app_id,
+                                   arg_parent_window,
+                                   options,
+                                   NULL,
+                                   start_done,
+                                   g_object_ref (request));
+
+  xdp_screen_cast_complete_start (object, invocation, request->id);
+
+  return TRUE;
+}
+
+static gboolean
+handle_take_pipewire_remote (XdpScreenCast *object,
+                             GDBusMethodInvocation *invocation,
+                             GUnixFDList *fd_list,
+                             const char *arg_session_handle,
+                             GVariant *arg_options)
+{
+  Call *call = call_from_invocation (invocation);
+  Session *session;
+  ScreenCastSession *screen_cast_session;
+  PipeWireRemote *remote;
+  int fd_id;
+  g_autoptr(GError) error = NULL;
+
+  session = acquire_session_from_call (arg_session_handle, call);
+  if (!session)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_ACCESS_DENIED,
+                                             "Invalid session");
+      return TRUE;
+    }
+
+  SESSION_AUTOLOCK_UNREF (session);
+
+  screen_cast_session = (ScreenCastSession *)session;
+
+  remote = screen_cast_session->pipewire_remote;
+  if (!remote)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_FAILED,
+                                             "No streams were started");
+      return TRUE;
+    }
+
+  fd_list = g_unix_fd_list_new ();
+  fd_id = g_unix_fd_list_append (fd_list,
+                                 pw_remote_steal_fd (remote->remote),
```

Comment:
I believe this leaks the fd. fd_list_append() dups the passed in fd and keeps that.